packages/fixie/src/agent.ts

```diff
@@ -317,30 +345,213 @@ export class FixieAgent {
+      console.log(`🌱 Agent stdout: ${sdata}`);
+    });
+    subProcess.stderr.on('data', (sdata: string) => {
+      console.error(`🌱 Agent stdout: ${sdata}`);
+    });
+    subProcess.on('close', (returnCode: number) => {
+      term(`🌱 Agent child process exited with code ${returnCode} - `).red('exiting.\n');
+    });
+    return subProcess;
+  }
+
+  /** Deploy an agent from the given directory. */
+  public static async DeployAgent(
+    client: FixieClient,
+    agentPath: string,
+    environmentVariables: Record<string, string> = {}
+  ): Promise<string> {
+    const config = await FixieAgent.LoadConfig(agentPath);
+    const agentId = `${(await client.userInfo()).username}/${config.handle}`;
+    term('🦊 Deploying agent ').green(agentId)('...\n');
+
+    // Check that the package.json path exists in this directory.
+    if (!fs.existsSync(path.resolve(path.join(agentPath, 'package.json')))) {
+      throw Error(`No package.json found in ${agentPath}. Only JS-based agents are supported.`);
+    }
+    const agent = await this.ensureAgent(client, agentId, config);
     const tarball = FixieAgent.getCodePackage(agentPath);
     const spinner = ora(' 🚀 Deploying...').start();
-    const revision = await agent.createRevision(tarball, environmentVariables);
+    const revision = await agent.createRevision({ tarball, environmentVariables });
     spinner.succeed(`Revision ${revision} was deployed to ${agent.agentUrl()}`);
     return revision;
   }
 
-  /** Get the current agent revision. */
-  public async getCurrentRevision(): Promise<AgentRevision> {
-    const result = await this.client.gqlClient().query({
-      fetchPolicy: 'no-cache',
-      query: gql`
-        query GetRevisionId($agentId: String!) {
-          agentById(agentId: $agentId) {
-            currentRevision {
-              id
-              created
+  /** Run an agent locally from the given directory. */
+  public static async ServeAgent({
+    client,
+    agentPath,
+    tunnel,
+    port,
+    environmentVariables,
+    debug,
+  }: {
+    client: FixieClient;
+    agentPath: string;
+    tunnel?: boolean;
+    port: number;
+    environmentVariables: Record<string, string>;
+    debug?: boolean;
+  }) {
+    const config = await FixieAgent.LoadConfig(agentPath);
+    const agentId = `${(await client.userInfo()).username}/${config.handle}`;
+    term('🦊 Serving agent ').green(agentId)('...\n');
+
+    // Check if the package.json path exists in this directory.
+    if (!fs.existsSync(`${agentPath}/package.json`)) {
+      throw Error(`No package.json found in ${agentPath}. Only JS-based agents are supported.`);
+    }
+
+    // Start the agent process locally.
+    const agentProcess = FixieAgent.spawnAgentProcess(agentPath, port);
+
+    // Wait for 5 seconds for it to start up.
+    await new Promise((resolve) => setTimeout(resolve, 5000));
+
+    async function* spawnTunnel(port: number): AsyncGenerator<string> {
```

Comment:
The promise/iterator/event emitter combo is really awkward here, but I don't understand the overall intent well enough to suggest an alternative. I think https://github.com/fixie-ai/ai-jsx/pull/271/files#r1321994681 will help me get a better handle on it.

Or we could say "we're a startup and we write scrappy code" and just yolo this if it works.